# Coding Exercises (Part 5)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## DataFrame Basics III

### Exercise 7: Sorting, Ranking & Summary Statistics

In [1]:
# run the cell!
import pandas as pd

In [2]:
# run the cell!
cars = pd.read_csv("cars.csv")

In [3]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

53. __Sort__ the __cars__ DataFrame by the __"mpg"__ column from __high to low__ and __save__ the new order!

In [4]:
cars.sort_values(by = "mpg", ascending = False, inplace= True)

__Inspect__ cars! What is the __name__ of the __most fuel efficient__ car?

In [5]:
#run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
322  46.6          4          86.0        65.0    2110          17.9   
329  44.6          4          91.0        67.0    1850          13.8   
325  44.3          4          90.0        48.0    2085          21.7   
394  44.0          4          97.0        52.0    2130          24.6   
326  43.4          4          90.0        48.0    2335          23.7   

     model_year  origin                  name  
322          80   japan             mazda glc  
329          80   japan   honda civic 1500 gl  
325          80  europe  vw rabbit c (diesel)  
394          82  europe             vw pickup  
326          80  europe    vw dasher (diesel)

The most fuel efficient car is ... mazda glc!

54. __Sort__ the __cars__ DataFrame by the __index__ from __low to high__. __Save__ the order!

In [6]:
cars.sort_index(ascending = True, inplace= True)

In [7]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

55. __Create__ a new column __"hp_rank"__ with the __horsepower rank__ for each car (__rank 1__ means __highest__ horsepower)! __Fill in the gaps!__

In [8]:
cars["hp_rank"] = cars.horsepower.rank(method = "min", ascending = False)

56. __Sort cars__ by the __"horsepower"__ from __high to low__. The __"hp_rank"__ for the car __"pontiac grand prix"__ is...? 

In [9]:
cars.sort_values(by = "horsepower", ascending = False)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
116  16.0          8         400.0       230.0    4278           9.5   
95   12.0          8         455.0       225.0    4951          11.0   
13   14.0          8         455.0       225.0    3086          10.0   
8    14.0          8         455.0       225.0    4425          10.0   
6    14.0          8         454.0       220.0    4354           9.0   
94   13.0          8         440.0       215.0    4735          11.0   
7    14.0          8         440.0       215.0    4312           8.5   
25   10.0          8         360.0       215.0    4615          14.0   
27   11.0          8         318.0       210.0    4382          13.5   
67   11.0          8         429.0       208.0    4633          11.0   
26   10.0          8         307.0       200.0    4376          15.0   
5    15.0          8         429.0       198.0    4341          10.0   
90   12.0          8         429.0       198.0    4952          11.5   
28    9.0          8         304.0       193.0    4732          18.5   
9    15.0          8         390.0       190.0    3850           8.5   
231  15.5          8         400.0       190.0    4325          12.2   
70   13.0          8         400.0       190.0    4422          12.5   
106  12.0          8         350.0       180.0    4499          12.5   
42   12.0          8         383.0       180.0    4955          11.5   
229  16.0          8         400.0       180.0    4220          11.1   
124  11.0          8         350.0       180.0    3664          11.0   
212  16.5          8         350.0       180.0    4380          12.1   
63   14.0          8         400.0       175.0    4385          12.0   
85   13.0          8         350.0       175.0    4100          13.0   
44   13.0          8         400.0       175.0    5140          12.0   
39   14.0          8         400.0       175.0    4464          11.5   
96   13.0          8         360.0       175.0    3821          11.0   
10   15.0          8         383.0       170.0    3563          10.0   
43   13.0          8         400.0       170.0    4746          12.0   
156  16.0          8         400.0       170.0    4668          11.5   
..    ...        ...           ...         ...     ...           ...   
378  38.0          4         105.0        63.0    2125          14.7   
348  37.7          4          89.0        62.0    2050          17.3   
332  29.8          4          89.0        62.0    1845          15.3   
145  32.0          4          83.0        61.0    2003          19.0   
310  38.1          4          89.0        60.0    1968          18.8   
345  35.1          4          81.0        60.0    1760          16.1   
248  36.1          4          91.0        60.0    1800          16.4   
55   27.0          4          97.0        60.0    1834          19.0   
196  24.5          4          98.0        60.0    2164          22.1   
343  39.1          4          79.0        58.0    1755          16.9   
218  36.0          4          79.0        58.0    1825          18.6   
59   23.0          4          97.0        54.0    2254          23.5   
181  33.0          4          91.0        53.0    1795          17.5   
198  33.0          4          91.0        53.0    1795          17.4   
144  31.0          4          76.0        52.0    1649          16.5   
246  32.8          4          78.0        52.0    1985          19.4   
195  29.0          4          85.0        52.0    2035          22.2   
394  44.0          4          97.0        52.0    2130          24.6   
117  29.0          4          68.0        49.0    1867          19.5   
244  43.1          4          90.0        48.0    1985          21.5   
326  43.4          4          90.0        48.0    2335          23.7   
325  44.3          4          90.0        48.0    2085          21.7   
102  26.0          4          97.0        46.0    1950          21.0   
19   26.0          4          97.0        46.0

The __"hp_rank"__ for the car __"pontiac grand prix"__ is... 1!

57. Check the amount of __unique values__ in __all columns__ of the cars DataFrame! __How many__ unique values are in the __"name"__ column?

In [10]:
cars.nunique()

mpg             129
cylinders         5
displacement     82
horsepower       93
weight          351
acceleration     95
model_year       13
origin            3
name            300
hp_rank          93
dtype: int64

There are ... 300 unique values in the "name" column! (There must be 98 name duplicates!)

58. __Select__ the __10__ cars with the __highest weights__!

In [11]:
cars.nlargest(n = 10, columns = "weight")

mpg  cylinders  displacement  horsepower  weight  acceleration  \
44   13.0          8         400.0       175.0    5140          12.0   
103  11.0          8         400.0       150.0    4997          14.0   
42   12.0          8         383.0       180.0    4955          11.5   
90   12.0          8         429.0       198.0    4952          11.5   
95   12.0          8         455.0       225.0    4951          11.0   
104  12.0          8         400.0       167.0    4906          12.5   
43   13.0          8         400.0       170.0    4746          12.0   
94   13.0          8         440.0       215.0    4735          11.0   
28    9.0          8         304.0       193.0    4732          18.5   
137  13.0          8         350.0       150.0    4699          14.5   

     model_year origin                          name  hp_rank  
44           71    usa           pontiac safari (sw)     23.0  
103          73    usa              chevrolet impala     46.0  
42           71    usa             dodge monaco (sw)     18.0  
90           73    usa      mercury marquis brougham     12.0  
95           73    usa      buick electra 225 custom      2.0  
104          73    usa                  ford country     33.0  
43           71    usa      ford country squire (sw)     28.0  
94           73    usa  chrysler new yorker brougham      6.0  
28           70    usa                      ih 1200d     14.0  
137          74    usa      buick century luxus (sw)     46.0

59. Call an appropriate method on the __cars__ DataFrame that provides plenty of __summary statistics__!

In [12]:
cars.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year     hp_rank  
count    398.000000  398.000000  392.000000  
mean      15.568090   76.010050  192.239796  
std        2.757689    3.697627  113.441056  
min        8.000000   70.000000    1.000000  
25%       13.825000   73.000000   98.500000  
50%       15.500000   76.000000  196.500000  
75%       17.175000   79.000000  294.000000  
max       24.800000   82.000000  391.000000

60. Create a __correlation matrix__ for the __cars__ DataFrame! Which factor has the __highest negative correlation__ to __mpg__?

In [13]:
cars.corr()

mpg  cylinders  displacement  horsepower    weight  \
mpg           1.000000  -0.775396     -0.804203   -0.778427 -0.831741   
cylinders    -0.775396   1.000000      0.950721    0.842983  0.896017   
displacement -0.804203   0.950721      1.000000    0.897257  0.932824   
horsepower   -0.778427   0.842983      0.897257    1.000000  0.864538   
weight       -0.831741   0.896017      0.932824    0.864538  1.000000   
acceleration  0.420289  -0.505419     -0.543684   -0.689196 -0.417457   
model_year    0.579267  -0.348746     -0.370164   -0.416361 -0.306564   
hp_rank       0.836399  -0.827170     -0.855328   -0.928239 -0.865003   

              acceleration  model_year   hp_rank  
mpg               0.420289    0.579267  0.836399  
cylinders        -0.505419   -0.348746 -0.827170  
displacement     -0.543684   -0.370164 -0.855328  
horsepower       -0.689196   -0.416361 -0.928239  
weight           -0.417457   -0.306564 -0.865003  
acceleration      1.000000    0.288137  0.653890  
model_year        0.288137    1.000000  0.390590  
hp_rank           0.653890    0.390590  1.000000

The factor weight shows the highest negative correlation to mpg! The lower the weight, the higher the fuel efficiency!

### Exercise 8: apply(), MultiIndex and Strings

In [14]:
# run the cell!
import pandas as pd

In [15]:
# run the cell!
cars = pd.read_csv("cars.csv")

In [16]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

61. __Calculate__ for all __numerical columns__ the __range between highest and lowest__ value! __Fill in the gaps!__ The range for __mpg__ is...?

In [17]:
cars.iloc[:, :-2].apply(lambda x: x.max() - x.min(), axis = 0)

mpg               37.6
cylinders          5.0
displacement     387.0
horsepower       184.0
weight          3527.0
acceleration      16.8
model_year        12.0
dtype: float64

The range for mpg is ... 37.6!

62. Create a __MultiIndex__ for the __cars__ DataFrame with the columns __"model_year"__ (outer level) and __"origin"__ (inner level). __Save__ the change!

In [18]:
cars.set_index(["model_year", "origin"], inplace = True)

63. __Sort__ both index levels in __ascending__ order!. __Save__ the change!

In [19]:
cars.sort_index(ascending = True, inplace = True )

__Inspect__ Cars! What is the __name__ of the __second__ car?

In [20]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
70         europe  26.0          4          97.0        46.0    1835   
           europe  25.0          4         110.0        87.0    2672   
           europe  24.0          4         107.0        90.0    2430   
           europe  25.0          4         104.0        95.0    2375   
           europe  26.0          4         121.0       113.0    2234   

                   acceleration                  name  
model_year origin                                      
70         europe          20.5  vw 1131 deluxe sedan  
           europe          17.5           peugeot 504  
           europe          14.5           audi 100 ls  
           europe          17.5              saab 99e  
           europe          12.5              bmw 2002

The name of the second car is ... peugeot 504!

64. __Select__ all cars with the "model_year" __75__!

In [21]:
cars.loc[75]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
origin                                                                    
europe  25.0          4          90.0        71.0    2223          16.5   
europe  29.0          4          90.0        70.0    1937          14.0   
europe  23.0          4         115.0        95.0    2694          15.0   
europe  23.0          4         120.0        88.0    2957          17.0   
europe  22.0          4         121.0        98.0    2945          14.5   
europe  25.0          4         121.0       115.0    2671          13.5   
japan   29.0          4          97.0        75.0    2171          16.0   
japan   24.0          4         134.0        96.0    2702          13.5   
japan   24.0          4         119.0        97.0    2545          17.0   
japan   33.0          4          91.0        53.0    1795          17.5   
usa     19.0          6         225.0        95.0    3264          16.0   
usa     18.0          6         250.0       105.0    3459          16.0   
usa     15.0          6         250.0        72.0    3432          21.0   
usa     15.0          6         250.0        72.0    3158          19.5   
usa     16.0          8         400.0       170.0    4668          11.5   
usa     15.0          8         350.0       145.0    4440          14.0   
usa     16.0          8         318.0       150.0    4498          14.5   
usa     14.0          8         351.0       148.0    4657          13.5   
usa     17.0          6         231.0       110.0    3907          21.0   
usa     16.0          6         250.0       105.0    3897          18.5   
usa     15.0          6         258.0       110.0    3730          19.0   
usa     18.0          6         225.0        95.0    3785          19.0   
usa     21.0          6         231.0       110.0    3039          15.0   
usa     20.0          8         262.0       110.0    3221          13.5   
usa     13.0          8         302.0       129.0    3169          12.0   
usa     23.0          4         140.0        83.0    2639          17.0   
usa     20.0          6         232.0       100.0    2914          16.0   
usa     23.0          4         140.0        78.0    2592          18.5   
usa     18.0          6         171.0        97.0    2984          14.5   
usa     19.0          6         232.0        90.0    3211          17.0   

                             name  
origin                             
europe                  vw dasher  
europe                  vw rabbit  
europe                 audi 100ls  
europe                peugeot 504  
europe                volvo 244dl  
europe                  saab 99le  
japan              toyota corolla  
japan               toyota corona  
japan                  datsun 710  
japan            honda civic cvcc  
usa       plymouth valiant custom  
usa                chevrolet nova  
usa               mercury monarch  
usa                 ford maverick  
usa              pontiac catalina  
usa             chevrolet bel air  
usa           plymouth grand fury  
usa                      ford ltd  
usa                 buick century  
usa     chevrolet chevelle malibu  
usa                   amc matador  
usa                 plymouth fury  
usa                 buick skyhawk  
usa           chevrolet monza 2+2  
usa               ford mustang ii  
usa                    ford pinto  
usa                   amc gremlin  
usa                 pontiac astro  
usa                    ford pinto  
usa                     amc pacer

65. __Select__ all cars with the "model_year" __76__ from __japan__! __How many__ cars do we get?

In [22]:
cars.loc[(76, "japan")]

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
76         japan   33.0          4          91.0        53.0    1795   
           japan   32.0          4          85.0        70.0    1990   
           japan   28.0          4          97.0        75.0    2155   
           japan   19.0          6         156.0       108.0    2930   

                   acceleration            name  
model_year origin                                
76         japan           17.4     honda civic  
           japan           17.0    datsun b-210  
           japan           16.4  toyota corolla  
           japan           15.5  toyota mark ii

There are ... 4 cars!

66. __Select__ all cars with the "model_year" __77__ and __79__  from __japan__! And only the columns __"mpg"__ and __"name"__! What is the __mpg__ of __datsun 210__?

In [23]:
cars.loc[([77,79], "japan"), ["mpg", "name"]]

mpg                     name
model_year origin                               
77         japan   31.5        honda accord cvcc
           japan   33.5    datsun f-10 hatchback
           japan   26.0  toyota corolla liftback
           japan   30.0                subaru dl
           japan   22.0               datsun 810
           japan   21.5               mazda rx-4
79         japan   34.1         mazda glc deluxe
           japan   31.8               datsun 210

The mpg of datsun 210 is ... 31.8!

67. __Swap the levels__ of the MultiIndex and then __sort__ the new MultiIndex in __ascending__ order! __Reassign cars__! __Fill in the gaps__!

In [24]:
cars = cars.swaplevel().sort_index(ascending = True)

In [25]:
# run the cell!
cars.head()

mpg  cylinders  displacement  horsepower  weight  \
origin model_year                                                      
europe 70          26.0          4          97.0        46.0    1835   
       70          25.0          4         110.0        87.0    2672   
       70          24.0          4         107.0        90.0    2430   
       70          25.0          4         104.0        95.0    2375   
       70          26.0          4         121.0       113.0    2234   

                   acceleration                  name  
origin model_year                                      
europe 70                  20.5  vw 1131 deluxe sedan  
       70                  17.5           peugeot 504  
       70                  14.5           audi 100 ls  
       70                  17.5              saab 99e  
       70                  12.5              bmw 2002

68. __Reset__ the __index__ of __cars__ and create a RangeIndex! __Save__ the change!

In [26]:
cars.reset_index(inplace = True)

In [27]:
# run the cell!
cars.head()

origin  model_year   mpg  cylinders  displacement  horsepower  weight  \
0  europe          70  26.0          4          97.0        46.0    1835   
1  europe          70  25.0          4         110.0        87.0    2672   
2  europe          70  24.0          4         107.0        90.0    2430   
3  europe          70  25.0          4         104.0        95.0    2375   
4  europe          70  26.0          4         121.0       113.0    2234   

   acceleration                  name  
0          20.5  vw 1131 deluxe sedan  
1          17.5           peugeot 504  
2          14.5           audi 100 ls  
3          17.5              saab 99e  
4          12.5              bmw 2002

69. __Convert__ all elements in the __origin__ column to __titlecase__! __Overwrite__ the __origin__ column!

In [28]:
cars.origin = cars.origin.str.title()

In [29]:
# run the cell!
cars.head()

origin  model_year   mpg  cylinders  displacement  horsepower  weight  \
0  Europe          70  26.0          4          97.0        46.0    1835   
1  Europe          70  25.0          4         110.0        87.0    2672   
2  Europe          70  24.0          4         107.0        90.0    2430   
3  Europe          70  25.0          4         104.0        95.0    2375   
4  Europe          70  26.0          4         121.0       113.0    2234   

   acceleration                  name  
0          20.5  vw 1131 deluxe sedan  
1          17.5           peugeot 504  
2          14.5           audi 100 ls  
3          17.5              saab 99e  
4          12.5              bmw 2002

Now, we want to __extract the manufacturer names__ from the column __"name"__. <br>
70. __Split__ the strings in the column __"name"__ on the very __first whitespace__ and __expand__ the splitted strings into __separate columns__! __Save__ new DataFrame in the variable __split__! __Fill in the gaps!__

In [30]:
split = cars.name.str.split(pat = " ", n = 1, expand = True)

__Inspect__!

In [31]:
# run the cell!
split.head()

0                  1
0       vw  1131 deluxe sedan
1  peugeot                504
2     audi             100 ls
3     saab                99e
4      bmw               2002

71. __Select__ the __first column__ and determine the __frequency of unique values__. __Fill in the gaps__! Which manufacturer is __most frequent__?

In [32]:
split[0].value_counts()

ford             51
chevrolet        47
plymouth         31
dodge            28
amc              28
toyota           26
datsun           23
vw               22
buick            17
pontiac          16
honda            13
mazda            12
mercury          12
oldsmobile       10
peugeot           8
fiat              8
audi              7
volvo             6
chrysler          6
renault           5
opel              4
subaru            4
saab              4
mercedes-benz     3
bmw               2
cadillac          2
triumph           1
nissan            1
ih                1
Name: 0, dtype: int64

There are 51 cars from the manufacturer ... Ford!

72. Let´s __doublecheck__ the result from above. Check __how many__ car names in the column "names" __contain "ford"__!

In [33]:
cars.name.str.contains("ford").sum()

51

It´s ... again 51! It seems that the very first word in the column "name" is always the manufacturer name.

# Well Done!

------------------------------

# Hints (Spoiler!)

53. sort_values() method

54. sort_index() method

55. rank() method, descending order

56. sort_values() method

57. nunique() method

58. nlargest() method

59. describe() method

60. corr() method

61. apply() method, maximum - minimum

62. set_index() method, pass list with column labels

63. sort_index() method

64. pass outer index labels to loc operator

65. cars.loc[(outer index labels, inner index labels)]

66. cars.loc[(outer index labels, inner index labels), column labels]

67. methods swaplevel() and sort_index()

68. reset_index() method

69. string(str)-method title()

70. string(str)-method split(), n = 1, 

71. value_counts() method 

72. string(str)-method contains()